crates/nu-command/src/network/http/client.rs

```diff
@@ -221,132 +221,168 @@ pub fn send_request(
+        _ => {
+            return Err(ShellErrorOrRequestError::ShellError(ShellError::IOError {
+                msg: "unsupported body input for JSON body type".into(),
+            }))
+        }
+    };
+    send_cancellable_request(request_url, Box::new(|| req.send_json(data)), span, signals)
+}
 
-                    let request_fn = move || {
-                        // coerce `data` into a shape that send_form() is happy with
-                        let data = data
-                            .iter()
-                            .map(|(a, b)| (a.as_str(), b.as_str()))
-                            .collect::<Vec<(&str, &str)>>();
-                        req.send_form(&data)
-                    };
-                    send_cancellable_request(&request_url, Box::new(request_fn), span, signals)
-                }
-                // multipart form upload
-                Value::Record { val, .. } if body_type == BodyType::Multipart => {
-                    let mut builder = MultipartWriter::new();
-
-                    let err = |e| {
-                        ShellErrorOrRequestError::ShellError(ShellError::IOError {
-                            msg: format!("failed to build multipart data: {}", e),
-                        })
-                    };
-
-                    for (col, val) in val.into_owned() {
-                        if let Value::Binary { val, .. } = val {
-                            let headers = [
-                                "Content-Type: application/octet-stream".to_string(),
-                                "Content-Transfer-Encoding: binary".to_string(),
-                                format!(
-                                    "Content-Disposition: form-data; name=\"{}\"; filename=\"{}\"",
-                                    col, col
-                                ),
-                                format!("Content-Length: {}", val.len()),
-                            ];
-                            builder
-                                .add(&mut Cursor::new(val), &headers.join("\n"))
-                                .map_err(err)?;
-                        } else {
-                            let headers =
-                                format!(r#"Content-Disposition: form-data; name="{}""#, col);
-                            builder
-                                .add(val.coerce_into_string()?.as_bytes(), &headers)
-                                .map_err(err)?;
-                        }
-                    }
-                    builder.finish();
+fn send_form_request(
+    request_url: &str,
+    body: Value,
+    req: Request,
+    span: Span,
+    signals: &Signals,
+) -> Result<Response, ShellErrorOrRequestError> {
+    match body {
+        Value::List { vals, .. } => {
+            if vals.len() % 2 != 0 {
+                return Err(ShellErrorOrRequestError::ShellError(ShellError::IOError {
+                    msg: "unsupported body input".into(),
+                }));
```

Comment:
Certainly a stretch goal but a lot of those errors are a bit generic and don't inform on the requirements either on the type or structure of the data. And labelling them as `ShellError::IOError` seems misplaced if the failure is actually not in the IO part.

Reply:
This certainly can be better, I'll take a look.

Reply:
This certainly can be better, I'll take a look.